### PR TITLE
Fix plugin name in fsgroupapplymetrics of csi and flexvolume

### PR DESCRIPTION
**Pull request by JornShen**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

change plugin name in fsgroupapplymetrics of csi and flexvolume to distinguish different driver

**Which issue(s) this PR fixes**:

Fixes  https://github.com/kubernetes/kubernetes/pull/95866#pullrequestreview-517225836

**Special notes for your reviewer**:

/cc @gnufied @jsafrane @mattcary @msau42

**Does this PR introduce a user-facing change?**:

```release-note
change plugin name in fsgroupapplymetrics of csi and flexvolume to distinguish different driver
```

### CODE REVIEW

**Comment by jsafrane** [comment]
IMO it's better to use `GetFullQualifiedPluginNameForVolume` to get the same plugin name as other operations captured by the same metric (mount / unmount / attach / detach / ...).

**Comment by JornShen** [comment]
> IMO it's better to use `GetFullQualifiedPluginNameForVolume` to get the same plugin name as other operations captured by the same metric (mount / unmount / attach / detach / ...).

@jsafrane 

`GetFullQualifiedPluginNameForVolume` seems that get driver name only from `CSIPersistentVolumeSource`.

https://github.com/kubernetes/kubernetes/blob/87cc9bca0fcbbdf1b8232c9c71e50023d62667c3/pkg/volume/util/metrics.go#L122-L127

while  I see drivername  in csi_plugin get from `CSIVolumeSource` and `CSIPersistentVolumeSource`

https://github.com/kubernetes/kubernetes/blob/87cc9bca0fcbbdf1b8232c9c71e50023d62667c3/pkg/volume/csi/csi_plugin.go#L357-L370

Is here still suitable to use `GetFullQualifiedPluginNameForVolume` to produce `fullPluginName`?

**Comment by jsafrane** [comment]
Yes, please use `GetFullQualifiedPluginNameForVolume`, as it's used in all other metrics, for example here: https://github.com/kubernetes/kubernetes/blob/b8c0435bc2b71775a7bd5056136b09575815696d/pkg/volume/util/operationexecutor/operation_generator.go#L698

You can fix `GetFullQualifiedPluginNameForVolume` to report correct CSI driver name for in-line CSI volumes, that would fix all  metrics.

**Comment by JornShen** [comment]
> Yes, please use `GetFullQualifiedPluginNameForVolume`, as it's used in all other metrics, for example here:
> 
> https://github.com/kubernetes/kubernetes/blob/b8c0435bc2b71775a7bd5056136b09575815696d/pkg/volume/util/operationexecutor/operation_generator.go#L698
> 
> You can fix `GetFullQualifiedPluginNameForVolume` to report correct CSI driver name for in-line CSI volumes, that would fix all metrics.

All right. I will try to add some unit test, and take some time.

File: pkg/volume/flexvolume/mounter.go [hunk 1]
```diff
@@ -94,7 +95,9 @@ func (f *flexVolumeMounter) SetUpAt(dir string, mounterArgs volume.MounterArgs)
 
 	if !f.readOnly {
 		if f.plugin.capabilities.FSGroup {
-			volume.SetVolumeOwnership(f, mounterArgs.FsGroup, mounterArgs.FSGroupChangePolicy, util.FSGroupCompleteHook(f.plugin.GetPluginName()))
+			// fullPluginName helps to distinguish different driver from flex volume plugin
+			fullPluginName := fmt.Sprintf("%s:%s", flexVolumePluginName, f.driverName)
```
Comment (JornShen) [inline pkg/volume/flexvolume/mounter.go:99]: @msau42 for flex volume, is it ok ?  The format is same as csi plugin.

Comment (msau42) [reply pkg/volume/flexvolume/mounter.go:99]: hmmm it seems like the flexvolume name is already qualified with the specific driver. I see metrics formatted with the "flexvolume-" prefix, although I haven't been able to find where in the code that prefix gets added yet.

File: pkg/volume/csi/csi_mounter.go [hunk 2]
```diff
@@ -279,7 +279,9 @@ func (c *csiMountMgr) SetUpAt(dir string, mounterArgs volume.MounterArgs) error
 	}
 
 	if c.supportsFSGroup(fsType, mounterArgs.FsGroup, c.fsGroupPolicy) {
-		err := volume.SetVolumeOwnership(c, mounterArgs.FsGroup, mounterArgs.FSGroupChangePolicy, util.FSGroupCompleteHook(c.plugin.GetPluginName()))
+		// fullPluginName helps to distinguish different driver from csi plugin
+		fullPluginName := fmt.Sprintf("%s:%s", c.plugin.GetPluginName(), c.driverName)
+		err := volume.SetVolumeOwnership(c, mounterArgs.FsGroup, mounterArgs.FSGroupChangePolicy, util.FSGroupCompleteHook(fullPluginName))
 		if err != nil {
```
Comment (gnufied) [inline pkg/volume/csi/csi_mounter.go:285]: Can we fix `GetFullQualifiedPluginNameForVolume` instead to consider `CSIVolumeSource` correctly?

File: pkg/volume/flexvolume/mounter.go (hunk 1 above)
Comment (gnufied) [reply pkg/volume/flexvolume/mounter.go:99]: I haven't dug too deeply but driver name includes the prefix from the way plugins are probed. we could just use `drivername` here.

Comment (jsafrane) [reply pkg/volume/flexvolume/mounter.go:99]: Other flex volume metrics use `GetFullQualifiedPluginNameForVolume` too, so I'd prefer to use it in SetVolumeOwnership too.

File: pkg/volume/csi/csi_mounter.go (hunk 2 above)
Comment (JornShen) [reply pkg/volume/csi/csi_mounter.go:284]: updated.

**Comment by JornShen** [comment]
I add some unit test in pkg/volume/util/metrics_test.go.
However meet some problems. :/

I used to use function `GatherAndCompare` of  `testutil` from "k8s.io/component-base/metrics/testutil".
But it has lint problem when it judge our metrics. And error as fellow:
```
=== RUN   TestOperationCompleteHook
--- FAIL: TestOperationCompleteHook (255.82s)
=== RUN   TestOperationCompleteHook/#00
    metrics_test.go:83: lint error: storage_operation_status_count:counter metrics should have "_total" suffix,storage_operation_status_count:non-histogram and non-summary metrics should not have "_count" suffix
    --- FAIL: TestOperationCompleteHook/#00 (253.74s)
=== RUN   TestOperationCompleteHook/#01
    metrics_test.go:83: lint error: storage_operation_status_count:counter metrics should have "_total" suffix,storage_operation_status_count:non-histogram and non-summary metrics should not have "_count" suffix
    --- FAIL: TestOperationCompleteHook/#01 (2.07s)
```

And error from the code as fellow:

https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/component-base/metrics/testutil/testutil.go#L48-L57

it add from MR, https://github.com/kubernetes/kubernetes/pull/92071.

Then I use the `GatherAndCompare`  from "github.com/prometheus/client_golang/prometheus/testutil".
But out CI tell us we it doesn't support using `prometheus/testutil` in `pkg/volume/util` in https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/95892/pull-kubernetes-bazel-test/1322015575634874368

So I change metrics name "storage_operation_status_count"  to "storage_operation_status_total" to fit standard metrics name rule.  Is it ok ? 

@gnufied @jsafrane @mattcary @msau42 

The changed name can also talk about. :)

**Comment by jsafrane** [comment]
I don't know enough about metric, @gnufied @msau42, can we rename a metric like this?

**Comment by msau42** [comment]
> can we rename a metric like this?

Given that this metric is alpha level, it is within bounds to have breaking changes. That being said, I also wanted to make one more breaking change to this metric, to combine both latency and error status into one metric, because right now, the latency metric only tracks latency for successful operations and not error operations. However, I don't think I'll have time to do this for 1.20, and would prefer not to make 2 breaking changes to the metric in a row.

For the dependency issue on prometheus/testutil, does running "hack/update-bazel.sh" fix the error? (try this in a different branch in case it doesn't work and we need to come back to this)

**Comment by JornShen** [comment]
> > can we rename a metric like this?
> 
> Given that this metric is alpha level, it is within bounds to have breaking changes. That being said, I also wanted to make one more breaking change to this metric, to combine both latency and error status into one metric, because right now, the latency metric only tracks latency for successful operations and not error operations. However, I don't think I'll have time to do this for 1.20, and would prefer not to make 2 breaking changes to the metric in a row.
> 
> For the dependency issue on prometheus/testutil, does running "hack/update-bazel.sh" fix the error? (try this in a different branch in case it doesn't work and we need to come back to this)

**Comment by JornShen** [comment]
> > can we rename a metric like this?
> 
> Given that this metric is alpha level, it is within bounds to have breaking changes. That being said, I also wanted to make one more breaking change to this metric, to combine both latency and error status into one metric, because right now, the latency metric only tracks latency for successful operations and not error operations. However, I don't think I'll have time to do this for 1.20, and would prefer not to make 2 breaking changes to the metric in a row.
> 
> For the dependency issue on prometheus/testutil, does running "hack/update-bazel.sh" fix the error? (try this in a different branch in case it doesn't work and we need to come back to this)

@msau42 

No, It is not the problem of "hack/update-bazel.sh". I have used prometheus/testutil before, including updated code of running "hack/update-bazel.sh". pkg `prometheus/testutil` have strict limit using place. 

So we can push the unit test for  `OperationCompleteHook` later (in suitable time of future with other change)?

**Comment by msau42** [comment]
> So we can push the unit test for `OperationCompleteHook` later (in suitable time of future with other changes)?

Yes, let's defer the unit test and other breaking metric changes to 1.21. If you are interested in working on this for the next release, that would be greatly appreciated!

**Comment by JornShen** [comment]
> > So we can push the unit test for `OperationCompleteHook` later (in suitable time of future with other changes)?
> 
> Yes, let's defer the unit test and other breaking metric changes to 1.21. If you are interested in working on this for the next release, that would be greatly appreciated!

All right.
updated.
I move out the unit test for `OperationCompleteHook `.

/recc @msau42  @jsafrane

**Comment by dashpole** [comment]
/assign @logicalhan 
for instrumentation

**Comment by msau42** [comment]
/lgtm
/approve

Thanks for quickly fixing this!

**Pull request merged**